### PR TITLE
Run non-optimization updates on predict for stateful RNN's

### PR DESCRIPTION
**Pull request by mynameisfiber**
Ideally, I could set my RNN's to stateful and use them to predict sequences in a stateful way.  Right now, the `_predict` function generated in the model does not run the updates necessary to set the RNN state.

~~The one problem I see is that we have collision with other layers that also set updates, namely the `BatchNormalization` layer.  A solution to this is to have the RNN state updates go to another variable, `state_updates` and leave `updates` for things that should only happen on training.  This seems like the right way of doing things, but the `get_param` function is thrown around a lot with tuple expansion and it seems like a mess to dig into that (maybe a good place for a `NamedTuple`).~~

EDIT: 56fa445 adds new property, `state_updates` to the `model` object that returns only the updates defined on layers set as stateful.  This allows us to update states on prediction only for layers that require this functionality to remain stateful.

### CODE REVIEW

File: keras/layers/containers.py [hunk 1]
```diff
@@ -75,6 +75,19 @@ def updates(self):
         return updates
 
     @property
+    def state_updates(self):
```
Comment (fchollet) [inline keras/layers/containers.py]: A docstring comment on how the system works would be welcome. I get it, but the 2 different behaviors make it quite non-obvious.

File: keras/layers/containers.py [hunk 2]
```diff
@@ -75,6 +75,19 @@ def updates(self):
...
+        state_updates = []
+        for l in self.layers:
+            if l.trainable and getattr(l, 'stateful', False):
```
Comment (fchollet) [inline keras/layers/containers.py]: Unclear why it should be restricted to trainable layers. Having this behavior in any stateful layer is fine imo.

**Comment by fchollet** [comment]
These are simple changes, but still, they should be tested. Let's add a few lines in `test_recurrent.py` to assert the stateful behavior of the RNNs in predict mode, and test the container-level `reset_states` method.

File: keras/layers/containers.py (hunk 2 above)
Comment (mynameisfiber) [inline keras/layers/containers.py]: To tell you the truth, I wasn't 100% sure what `trainable` was signaling and it seemed to indicate that a layer had been compiled and would accept data so it felt right...  I took it out though.

File: keras/layers/containers.py (hunk 1 above)
Comment (mynameisfiber) [inline keras/layers/containers.py]: Docstring requests will always be respected.

**Comment by mynameisfiber** [comment]
@fchollet Comments should be taken care of... I also added one more commit (bf8dc2f) that cleaned up the `_step` closure in the `rnn` function for the theano backend.
